scheds/rust/scx_layered/src/stats.rs

```diff
@@ -429,32 +426,101 @@ impl SysStats {
+#[derive(Debug)]
+pub enum StatsReq {
+    Hello(ThreadId),
+    Refresh(ThreadId, Stats),
+    Bye(ThreadId),
+}
+
+#[derive(Debug)]
+pub enum StatsRes {
+    Hello(Stats),
+    Refreshed((Stats, SysStats)),
+    Bye,
+}
+
+pub fn launch_server() -> Result<ScxStatsServer<StatsReq, StatsRes>> {
+    let open: Box<dyn StatsOpener<StatsReq, StatsRes>> = Box::new(move |(req_ch, res_ch)| {
+        let tid = current().id();
+        req_ch.send(StatsReq::Hello(tid))?;
+        let mut stats = Some(match res_ch.recv()? {
+            StatsRes::Hello(v) => v,
+            res => bail!("invalid response to Hello: {:?}", &res),
+        });
+
+        let read: Box<dyn StatsReader<StatsReq, StatsRes>> =
+            Box::new(move |_args, (req_ch, res_ch)| {
+                req_ch.send(StatsReq::Refresh(tid, stats.take().unwrap()))?;
+                let (new_stats, sys_stats) = match res_ch.recv()? {
+                    StatsRes::Refreshed(v) => v,
+                    res => bail!("invalid response to Refresh: {:?}", &res),
+                };
+                stats = Some(new_stats);
+                sys_stats.to_json()
+            });
+
+        Ok(read)
+    });
+
+    let close: Box<dyn StatsCloser<StatsReq, StatsRes>> = Box::new(move |(req_ch, res_ch)| {
+        req_ch.send(StatsReq::Bye(current().id())).unwrap();
+        match res_ch.recv().unwrap() {
+            StatsRes::Bye => {}
+            res => panic!("invalid response to Bye: {:?}", &res),
+        }
+    });
+
+    Ok(ScxStatsServer::new()
         .add_stats_meta(LayerStats::meta())
         .add_stats_meta(SysStats::meta())
-        .add_stats(
+        .add_stats_ops(
             "top",
-            Box::new(move |_| sys_stats.lock().unwrap().to_json()),
+            ScxStatsOps {
+                open,
+                close: Some(close),
+            },
         )
-        .launch()?;
-    Ok(())
+        .launch()?)
 }
 
-pub fn monitor(shutdown: Arc<AtomicBool>) -> Result<()> {
-    let mut client = ScxStatsClient::new().connect()?;
-    let mut last_at = 0.0;
-
+pub fn monitor(intv: Duration, shutdown: Arc<AtomicBool>) -> Result<()> {
+    let mut retry_cnt: u32 = 0;
     while !shutdown.load(Ordering::Relaxed) {
-        let sst = client.request::<SysStats>("stats", vec![])?;
-        if sst.at != last_at {
+        let mut client = match ScxStatsClient::new().connect() {
+            Ok(v) => v,
+            Err(e) => match e.downcast_ref::<std::io::Error>() {
+                Some(ioe) if ioe.kind() == std::io::ErrorKind::ConnectionRefused => {
+                    if retry_cnt == 1 {
+                        info!("Stats server not avaliable, retrying...");
+                    }
+                    retry_cnt += 1;
+                    sleep(Duration::from_secs(1));
+                    continue;
+                }
+                _ => Err(e)?,
+            },
+        };
+        retry_cnt = 0;
+
+        while !shutdown.load(Ordering::Relaxed) {
```

Comment:
is this inner one needed?

Reply:
I think so. Otherwise, the inner loop will never exit on ctrl-c.